File: python/bullcowComputer.py
#!/usr/bin/python3
import sys

guesses = []

def countBullsCows(secret, guess):
    bulls = 0
    cows = 0
    for i in range(4):
        if secret[i] == guess[i]:
            bulls += 1
        elif secret.find(guess[i]) != -1:
            cows += 1
    return bulls * 10 + cows

def generate():
    for v in range(100, 10000):
        s = f'{v:04d}'
        ok = True
        for i in range(4):
            if s.find(s[i], i + 1) != -1:
                ok = False
        for g in guesses:
            if countBullsCows(s, g['guess']) \
                != g['bulcow']:
                ok = False
        if ok:
            return s
    return "9999"

def readBullsCows():
    while True:
        bulls, cows = [int(x)
            for x in input('Введите число быков и число коров: ').split()]
        if bulls + cows >= 0 \
            and bulls + cows <= 4 \
            and bulls >=0 and cows >=0:
            return bulls * 10 + cows

def save(guess, bulcow):
    guesses.append({'guess': guess, 'bulcow': bulcow})

while True:
    guess = generate()
    print(f'Моя попытка: {guess}')
    bulcow = readBullsCows()
    if bulcow == 40:
        print('Я отгадал!')
        break
    save(guess, bulcow)

File: cpp/bullcowComputer.cpp
#include <string>
#include <iostream>
#include <list>

struct Bulcow {
    std::string guess;
    int bulcow;
};
std::list<Bulcow> guesses;

int countBullsCows(std::string secret, std::string guess)
{
    int bulls = 0;
    int cows = 0;
    for (int i = 0 ; i < 4 ; ++i)
    {
        if (guess[i] == secret[i])
        {
            ++bulls;
        }
        else if (secret.find(guess[i]) != std::string::npos)
        {
            ++cows;
        }
    }
    return bulls * 10 + cows;
}

bool check(std::string s)
{
    for (int i = 0 ; i < 4 ; ++i)
    {
        if (s.find(s[i], i + 1) != std::string::npos)
        {
            return false;
        }
    }
    return true;
}

std::string generate()
{
    for (int v = 100 ; v <= 9999 ; ++v)
    {
        std::string s = std::to_string(v);
        if (v < 1000)
        {
            s = "0" + s;
        }
        if (!check(s))
        {
            continue;
        }
        bool ok = true;
        for (const Bulcow &g : guesses)
        {
            if (countBullsCows(s, g.guess) != g.bulcow)
            {
                ok = false;
            }
        }
        if (ok)
        {
            return s;
        }
    }
    return "9999";
}

int readBullsCows()
{
    while (true)
    {
        int bulls, cows;
        std::cout << "Введите число быков и число коров: ";
        std::cin >> bulls >> cows;
        if (bulls + cows >= 0
            && bulls + cows <= 4
            && bulls >=0 && cows >=0)
        {
            return bulls * 10 + cows;
        }
    }
}

void save(std::string guess, int bulcow)
{
    guesses.push_back({guess, bulcow});
}

int main()
{
    while (true)
    {
        std::string guess = generate();
        std::cout << "Моя попытка: " << guess << "\n";
        int bulcow = readBullsCows();
        if (bulcow == 40)
        {
            std::cout << "Я отгадал!\n";
            break;
        }
        save(guess, bulcow);
    }
}
